jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -191,40 +202,128 @@ public int getMaxCapacity()
+
+        // Try to acquire a pooled entry.
         Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
+        if (entry != null)
+        {
+            bucket.recordPooled();
+            updateMemory(-bucket.getCapacity(), direct);
+            RetainableByteBuffer buffer = entry.getPooled();
+            ((Buffer)buffer).acquire();
+            return buffer;
+        }
+
+        return newRetainableByteBuffer(bucket.getCapacity(), direct, buffer -> reserve(bucket, buffer));
+    }
+
+    private void reserve(RetainedBucket bucket, RetainableByteBuffer buffer)
+    {
+        bucket.recordRelease();
+
+        boolean direct = buffer.isDirect();
+
+        // Discard the buffer if maxMemory is exceeded.
+        long excessMemory = getExcessMemory(bucket, direct);
+        if (excessMemory > 0)
+        {
+            bucket.recordNonPooled();
+            return;
+        }
 
-        RetainableByteBuffer buffer;
+        Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().reserve();
+        // Cannot reserve, discard the buffer.
         if (entry == null)
         {
-            Pool.Entry<RetainableByteBuffer> reservedEntry = bucket.getPool().reserve();
-            if (reservedEntry != null)
-            {
-                buffer = newRetainableByteBuffer(bucket._capacity, direct, retainedBuffer ->
-                {
-                    BufferUtil.reset(retainedBuffer.getByteBuffer());
-                    if (!reservedEntry.release())
-                        reservedEntry.remove();
-                });
+            bucket.recordNonPooled();
+            return;
+        }
 
-                // A reserved entry may become old and be evicted before it is enabled.
-                if (reservedEntry.enable(buffer, true))
-                {
-                    if (direct)
-                        _currentDirectMemory.addAndGet(buffer.capacity());
-                    else
-                        _currentHeapMemory.addAndGet(buffer.capacity());
-                    releaseExcessMemory(direct);
-                    return buffer;
-                }
+        ByteBuffer byteBuffer = buffer.getByteBuffer();
+        BufferUtil.reset(byteBuffer);
+        Buffer pooledBuffer = new Buffer(byteBuffer, b -> release(bucket, entry));
+        // Discard the buffer if the entry cannot be enabled.
+        if (entry.enable(pooledBuffer, false))
+        {
+            updateMemory(bucket.getCapacity(), direct);
+        }
+        else
+        {
+            bucket.recordNonPooled();
+            entry.remove();
+        }
+    }
+
+    private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> entry)
+    {
+        bucket.recordRelease();
+
+        RetainableByteBuffer buffer = entry.getPooled();
+        BufferUtil.reset(buffer.getByteBuffer());
+        boolean direct = buffer.isDirect();
+        long excessMemory = getExcessMemory(bucket, direct);
+        if (excessMemory > 0)
```

Comment:
This algorithm is racy and it cannot guarantee that the max capacity is respected.  Multiple threads might all calculate excess memory at the same time and all get an answer saying they are 1 byte below max.  So all  threads will not evict and all 10 will updateMemory below.

I think we should do an atomicUpdate here and return the excess from that.   We can then do the evicting and if that is not enough discard the buffer and update the memory again.

Reply:
@gregw turns out there always is a risk of over-shooting the max memory.

We can precisely update the memory when it's within maxMemory; but when the update would overshoot we would not update the memory, so another thread could see a stale value, because the first thread would now try to evict and free space.

Seems too complicated to try to be precise, so the current compromise is simple and speedy but not precise.

Reply:
@sbordet I know we cannot fix the race entirely.  But we can use the atomic that we already have to do a bit better than either having multiple threads not evicting when they should or multiple threads evicting when they should not.

The max memory is already atomic, so why not updating it atomically so that each thread knows its contribution to the amount over the max and then proceeds to try to evict only that much memory.

Even if it doesn't solve the race, it will be faster because we will not access the atomic twice (once to test and once to update).

Standby, I'll prepare a suggestion for this....